File: statistics/vector.ts
class Vector{
    data:Array<number>;
    /**
     * 向量
     * @param {Array<number>} ...data [向量数据]
     */
    constructor([...data]:Array<number>){
        this.data =data;
    }
    /**
     * 向量的和
     * @param  {vector}        data [求和向量]
     * @return {Array<number>}      [向量数据矩阵]
     */
    add(data:Vector):Array<number>{
        let tempAry:Array<number>=[];
        if(data instanceof Vector&& data.data.length==this.data.length){
            for(let i=0;i<data.data.length;i++){
                tempAry.push(this.data[i]+data.data[i])
            }
            return (this,tempAry);
        }
    }
    /**
     * 向量的差
     * @param  {vector}        data [向量]
     * @return {Array<Number>}      [返回向量数组]
     */
    mult(data:Vector):Array<Number>{
        let tempAry:Array<number> = [];
        if(data instanceof Vector && data.data.length==this.data.length){
            for(let i=0;i<data.data.length;i++){
                tempAry.push(data.data[i]-this.data[i]);
            }
            return (this,tempAry);
        }else{
            throw Error("参数必须是向量实例或两个向量长度需相同");
        }
    }
    /**
     * 对向量求模
     * @param  {Array<number>|Vector} data [求模的数据或向量]
     * @return {number}                    [结果]
     */
    mod(data:Array<number>|Vector):number{
        let total:number =0;
        if(data instanceof Array || data instanceof Vector){
        if(data instanceof Array){
            for(let i =0;i<data.length;i++){
                total+=data[i]**2
            }
            return (this,Math.sqrt(total));
        }
        if(data instanceof Vector){
            return this.mod(data.data);
        }else{
            throw TypeError("参数类型错误,参数必须是向量类或者数组类");
        }
    }
    }
    /**
     * 向量积
     * @param  {number|Array<number>|vector} data [求积的向量或者数组]
     * @return {number}                           [结果值]
     */
    product(data:number|Array<number>|Vector):number|Array<number>{
        let total:number =0;
        //倍数向量
        if(typeof data ==="number"){
            for(let i =0;i<this.data.length;i++){
                this.data[i]*=data;
            }
        }
        //乘积(点积)
        if(data instanceof Array){
            if(this.data.length==data.length){
                for(let i =0;i<data.length;i++){
                    total+=(this.data[i]*data[i]);
                }
                return (this,total);
            }else{
                throw TypeError("参数长度不一致!")
            }
        }
        //乘积(点积)
        if(data instanceof Vector){
            return (this,this.product(data.data));
        }
    }
    /**
     * 判断是否和另一个向量垂直,垂直返回真,不垂直返回假
     * @param  {vector|Array<number>} dada [和向量比较的向量或数据]
     * @return {boolean}                   [结果]
     */
    isVertical(dada:Vector|Array<number>):boolean{//点积的值为零返回真
        if(dada instanceof Array){
            if(this.product(dada)==0){
                return true;
            }
        }
        if(dada instanceof Vector){
            this.isVertical(dada.data)
        }
        return (this,false);
    }
    /**
     * [angle 向量夹角]
     * @param  {vector|Array<number>} data [向量或者向量数据]
     * @return {number}                    [结果值]
     */
    angle(data:Vector|Array<number>):number{
        if(data instanceof Array){
            var total:number =0;
           if(data.length==this.data.length){
            for(let i=0;i<data.length;i++){
                total+=(this.data[i]*data[i]);
           }
        }
            return (this,total/(this.mod(data)*this.mod(this.data)));
        }
        if(data instanceof Vector){
            this.angle(data.data);
        }
    }
}
export {Vector};

File: statistics/python/Vector.py
# -*- coding: utf-8 -*-
# @Author: Jingyuexing
# @Date:   2019-07-12 00:24:07
# @Last Modified by:   Jingyuexing
# @Last Modified time: 2020-01-12 17:20:47


class vector:
    """docstring for Vector
    向量
    """
    
    def __init__(self, data=[]):
        if isinstance(data, list):
            self.data = data
        else:
            print("Error:data type is not list")

    def add(self, data):
        '''向量相加

        [description]

        Arguments:
            data {Vector} -- 相加的向量

        Returns:
            list -- 向量和值
        '''
        tempAry = []
        if isinstance(data, Vector) and len(data.data) == len(self.data):
            for i in range(0, len(data.data)):
                tempAry.append(self.data[i] + data.data[i])
        return tempAry

    def mult(self, data):
        """[对向量两个向量进行求差]

        [description]

        Arguments:
            data {[Vector]} -- [相减向量]

        Returns:
            [list] -- [向量和结果值]
        """
        tempAry = []
        if isinstance(data, Vector) and len(data.data) == len(self.data):
            for i in range(0, len(data.data)):
                tempAry.append(data.data[i] - self.data[i])
            return tempAry
        else:
            print("Error:参数必须是向量实例或两个向量长度需相同")

    def mod(self, data):
        '''向量求模

        [description]

        Arguments:
            data {Vecotr} -- 向量

        Returns:
            number -- 模值
        '''
        total = 0
        if isinstance(data, Vector):
            for i in range(0, len(data.data)):
                total = total + data.data[i]**2
            return total**(1 / 2)
        else:
            return TypeError(data + "is not Vector")

    def product(self, data):
        '''向量乘积

        >>>Vector_1 = Vector([2,3,4,5,6,7])
        >>>Vector_2 = Vector([12,5,78,2,12])
        >>>Vector_2.product(Vector_1) return 748:number
        >>>Vector_2.product(8)  return [96, 40, 624, 16, 96, 360]
        Arguments:
            data {number | list | Vector} -- 求点积或者向量乘积

        Returns:
            number|list -- 向量乘积值
        '''
        total = 0
        if isinstance(data, int) or isinstance(data, float):
            for i in range(0, len(self.data)):
                self.data[i] = self.data[i] * data
        if isinstance(data, list):
            if len(self.data) == data.__len__():
                for i in range(0, len(data)):
                    total = total + self.data[i] * data[i]
                return total
            else:
                return TypeError("参数长度不一致!")
        if isinstance(data, Vector):
            return self.product(data.data)

    def isVertical(self, data):
        '''判断是否和另外的向量垂直,垂直返回真,不垂直返回假



        Arguments:
            data { Vector|List<number> } --  和自身比较的向量

        Returns:
            bool --   是否垂直
        '''
        if isinstance(data, list) and self.product(data) == 0:
            return True
        if isinstance(data, Vector):
            self.isVertical(data.data)
        return False

    def map(self):
        return self.data

    def angle(self, data):
        '''向量夹角
        Arguments:
            data {  Vector | list<number>} --   向量或者向量数据

        Returns:
            number -- 结果值
        '''
        if isinstance(data, list):
            total = 0
            if len(data) == len(self.data):
                for i in range(0, len(data)):
                    total = total + self.data[i] * data[i]
            return total / self.mod(data) * self.mod(self.data)
        if isinstance(data, Vector):
            self.angle(data.data)

    def insertData(self, *args):
        for i in args:
            self.data.append(i)
        return self
